Client/Application/AppLaunchUtil.swift

```diff
@@ -169,4 +187,56 @@ class AppLaunchUtil {
         // increase session count value
         profile.prefs.setInt(sessionCount + 1, forKey: PrefsKeys.SessionCount)
     }
+
+    private func runEarlyExperimentDependencies() {
+        runAppServicesHistoryMigration()
+    }
+
+    // MARK: - Application Services History Migration
+
+    private func runAppServicesHistoryMigration() {
+        let placesHistory = FxNimbus.shared.features.placesHistory.value()
+        FxNimbus.shared.features.placesHistory.recordExposure()
+        guard placesHistory.migration != .disabled else {
+            log.info("Migration disabled, won't run migration")
+            return
+        }
+        let browserProfile = self.profile as? BrowserProfile
+        let migrationRanKey = "PlacesHistoryMigrationRan" + placesHistory.migration.rawValue
+        let migrationRan = UserDefaults.standard.bool(forKey: migrationRanKey)
+        if !migrationRan {
+            log.info("Migrating Application services history")
+            let id = GleanMetrics.PlacesHistoryMigration.duration.start()
+            // We mark that the migration started
+            // this will help us identify how often the migration starts, but never ends
+            // additionally, we have a seperate metric for error rates
+            GleanMetrics.PlacesHistoryMigration.migrationEndedRate.addToNumerator(1)
+            GleanMetrics.PlacesHistoryMigration.migrationErrorRate.addToNumerator(1)
+            browserProfile?.migrateHistoryToPlaces(
+            migrationConfig: placesHistory.migration.into(),
+            callback: { result in
+                self.log.info("Successful Migration took \(result.totalDuration / 1000) seconds")
+                UserDefaults.standard.setValue(true, forKey: migrationRanKey)
+
+                // We record various success metrics here
+                GleanMetrics.PlacesHistoryMigration.duration.stopAndAccumulate(id)
+                GleanMetrics.PlacesHistoryMigration.numMigrated.set(Int64(result.numSucceeded))
+                self.log.info("Migrated \(result.numSucceeded) entries")
+                GleanMetrics.PlacesHistoryMigration.numToMigrate.set(Int64(result.numTotal))
+                GleanMetrics.PlacesHistoryMigration.migrationEndedRate.addToDenominator(1)
+            },
+            errCallback: { err in
+                let errMsg = err?.localizedDescription ?? "Unknown error during History migration"
```

Comment:
This will try again next start, right? For the dry-run, I wonder if we should store something like the number of times we tried to migrate and give up after a number of attempts? For a "real" migration it's a little trickier, but retrying indefinitely doesn't seem like the right thing in that case either, so I guess we'd just start those poor users with an empty db?

Reply:
Yup, it will keep trying until it succeeds - now that I think about it, I'm leaning more toward disabling the migration if it fails. or more generally, only attempting to run the migration once for the dry run. Because otherwise, users who keep failing for unrelated reasons (like corrupted db) will skew our data because they will keep on trying to migrate.

For real migration, we can retry a few times then have them start on an empty db